### DISCUSSION TITLE
About Mojo native support for disk io, networking, gui and threading

### DISCUSSION THREAD

**Discussion by lijh8** [modular/modular#1000] [General]
Will Mojo have native support for: 
- disk file io (fopen, fgets, ..), 
- networking (socket, epoll, boost.asio, python asyncio, ..), 
- gui (tkinter, gtkmm, Qt, ..), 
- multiple threading,
- portability across PC, Server, mobile, ..
- etc..

**Comment by ksandvik** [comment]
See 0.4 release notes where there's an initial file I/O module: https://docs.modular.com/mojo/changelog.html#v0.4.0-2023-10-05

**Comment by rarebreed** [comment]
Mojo is still very very young.  It doesn't even have a native dict/map type yet :) 

It would be nice if there was a list of functionality they want in their stdlib.  Will mojo be "batteries included" like python?  Or more lean like rust?  Will they have async forms from the beginning, or tack it on later like rust and python did?  This is a significant question, because in both rust's and python's cases, they have competing non-standard async versions of common modules.  In rust, tokio vs async-std, and in python, async vs trio vs anyio.

**Reply by lijh8** [reply]
How about make python a static typing language by introducing type inference. This will not break existing code. eg:
```
a = 123 # inferred as int
a = “hello” # re-defined as string
a:float = 123 # type annotation 
```

**Comment by ksandvik** [comment]
Interesting point. My take/opinion is single supported libraries based on open-source contributions/work. If someone then wants to make alternate libraries, that is fine, too, but it's outside the Mojo umbrella. Anyway, just an opinion.

**Reply by rarebreed** [reply]
My fear is that by not having an async "standard" module, you can make incompatible ecosystems where library foo uses my-async-network and lib foo has a transitive dependency that uses another-async-network lib and they are incompatible with each other.  Indeed, anyio is supposed to be able to handle plain old native async libraries like aiohttp, as well as trio style nurseries.

I also mention this, because in rust, there is a problem with async scoped concurrency that really didn't come to light earlier because they tacked on async later in language development.

**Comment by richardkiss** [comment]
See also Chris's response here https://github.com/modularml/mojo/discussions/1147
